### PR TITLE
BUG: Fix two bugs in `statsmodels.stats.power.normal_sample_size_one_tail`

### PR DESCRIPTION
**Pull request by goxberry**
- ~~[ ] closes #xxxx~~ (does not apply)
- [x] tests added / passed. 
- [x] code/documentation is well formatted.  
- [x] properly formatted commit message. See 
      [NumPy's guide](https://docs.scipy.org/doc/numpy-1.15.1/dev/gitwash/development_workflow.html#writing-the-commit-message). 

This pull request fixes two bugs in `statsmodels.stats.power.normal_sample_size_one_tail`:

1. If the default value of that method's `std_alternative` parameter is used, that method raises an exception because a `float` cannot be multiplied by a `NoneType`. This bug is mitigated by setting the `std_alternative` parameter to zero in the body of that method if its value is `None`.
2. The `power` parameter cannot be less than `alpha`. Admittedly, I don't know why someone would intentionally try to pass arguments satisfying that criterion, but nevertheless, a power less than alpha would be ill-posed. This bug is mitigated by adding a conditional checking this condition; if satisfied, a `ValueError` is raised. The method's documentation has also been updated.

Reviewing the code in `statsmodels.stats.power.normal_sample_size_one_tail` also suggests the implementation might be incorrect: after the mitigations above, assuming `std_alternative` is `None`, and all parameters other than `power` are fixed, the computed sample size will be the same for all valid values of `power`. One possible alternative implementation (omitting docstrings) could be something like:

```python
def normal_sample_size_one_tail(diff, power, alpha, std_null=1.,
                                 std_alternative=None):

    if power < alpha:
        raise ValueError(f"Power cannot be less than alpha; set power >= {alpha} to fix")

    if std_alternative is None:
        std_alternative = 0

    crit_power = stats.norm.isf(power)
    crit = stats.norm.isf(alpha)
    n1 = (std_null ** 2 + std_alternative ** 2) * (crit - crit_power)**2 / (diff ** 2)
    return n1
```
under the assumption that, in the two-sample case, the sample sizes are equal. This expression agrees with a similar expression for the two-sample case on page 490 in Chapter 10 of J. L. Devore, K. N. Berk, *Modern Mathematical Statistics with Applications*, 2nd Edition, in *Springer Texts on Statistics*, and when `std_alternative=0`, it agrees with the expression for the one-sample case on page 441 in Chapter 9 of the same textbook, as well as that on [Wikipedia's page for "sample size determination"](https://en.wikipedia.org/wiki/Sample_size_determination#Cumulative_distribution_function). For the one-tailed, one-sample case, this expression should be exact; for the two-tailed case, it makes the approximation of ignoring the smaller tail's contribution to Type II error. However, changing the implementation would affect some existing tests in `statsmodels/stats/tests/test_proportion.py`, so I have not changed the expression for `n1` in this pull request. If making the change above to the expression for  `n1` is sensible, I would be happy to contribute that change as well.

<details>
**Notes**:

* It is essential that you add a test when making code changes. Tests are not 
  needed for doc changes.
* When adding a new function, test values should usually be verified in another package (e.g., R/SAS/Stata).
* When fixing a bug, you must add a test that would produce the bug in main and
  then show that it is fixed with the new code.
* New code additions must be well formatted. Changes should pass flake8. If on Linux or OSX, you can
  verify you changes are well formatted by running 
  ```
  git diff upstream/main -u -- "*.py" | flake8 --diff --isolated
  ```
  assuming `flake8` is installed. This command is also available on Windows 
  using the Windows System for Linux once `flake8` is installed in the 
  local Linux environment. While passing this test is not required, it is good practice and it help 
  improve code quality in `statsmodels`.
* Docstring additions must render correctly, including escapes and LaTeX.

</details>

### CODE REVIEW

**Comment by josef-pkt** [comment]
Thanks, I will go through it

The function was not really written as a standalone function.
AFAICS, I wrote it mainly for use with proportion ztest

the intention or usecase for std_alternative is to allow std to differ between alternative and null.
E.g. in standard cases without inherent heteroscedasticity, like OLS/normal, the std would be the same under alternative and null

```
if std_alternative is None:
        std_alternative = std_null
```

second

`if power < alpha:` will break vectorization, which IIRC is allowed here.

**Comment by josef-pkt** [comment]
There is only one unit test for the function directly and that is in test_proportions

Other unit tests are for samplesize_proportions_2indep_onetail which delegates to normal_sample_size_one_tail

**Comment by goxberry** [comment]
> `if power < alpha:` will break vectorization, which IIRC is allowed here.

If you prefer, I could mask the vector and return `nan` for vector elements in which `power < alpha`. This approach is used in, e.g., `scipy.stats.rv_continuous` to indicate invalid input arguments for methods like `scipy.stats.norm.isf` and would not break vectorization.

**Comment by josef-pkt** [comment]
> If you prefer, I could mask the vector and return nan for vector elements in which power < alpha

Yes, but it might be better to do it on the return, because any of the arguments could be an array with broadcasting, but there is only one return scalar or array.

**Comment by goxberry** [comment]
> The function was not really written as a standalone function.
> AFAICS, I wrote it mainly for use with proportion ztest

I believe that's the only case under which this function would be valid. If the idea is to compute a sample size for a proportion z-test, then the current interface seems nonintuitive because standard deviations of proportions modeled as binomial distributions depend on their mean values, so the parameters `diff`, `std_null`, and `std_alternative` cannot vary independently of each other. All three of these quantities only depend on two parameters: the proportion used to formulate the alternative hypothesis (e.g., the `p_0` in `p < p_0`) and whatever proportion is used for a power calculation:

```python
def normal_sample_size_proportion(p_alt, p_power, alpha, power):

    # p_power = proportion used for power calculation
    crit_power = stats.norm.isf(power)
    crit = stats.norm.isf(alpha)
    diff = p_alt - p_power
    std_alt = numpy.sqrt(p_alt * (1 - p_alt))
    std_power = numpy.sqrt(p_power * (1 - p_power))
    n1 = (crit * std_alt - crit_power * std_power) ** 2 / (diff ** 2)
    return n1
```

> the intention or usecase for std_alternative is to allow std to differ between alternative and null.
> E.g. in standard cases without inherent heteroscedasticity, like OLS/normal, the std would be the same under alternative and null

As far as I can tell from the current formula, even if that were the case, it would under-count variance in an unpaired, nonproportion z-test by a factor of 2, and that choice would also restrict the sample size calculation to two-sample tests.

File: statsmodels/stats/power.py [hunk 1]
```diff
@@ -172,39 +172,48 @@ def normal_sample_size_one_tail(diff, power, alpha, std_null=1.,
...
+        sqrt(nobs)
 
     Returns
     -------
     nobs : float
         sample size to achieve the desired power
 
     """
+    if std_alternative is None:
+        std_alternative = 0
```
Comment (josef-pkt) [inline statsmodels/stats/power.py:204]: this should be = std_null

**Comment by josef-pkt** [comment]
I'm not remembering the details well enough

should diff be replaced by np.abs(diff) ?

(thinking more)
abs not needed: diff shows up only as `diff**2` (if `(crit * std_null - crit_power * std_alternative)` is positive)

what's "appropriate" tail?

> This uses an explicit computation for the sample size that is required
    to achieve a given power corresponding to the appropriate tails of the
    normal distribution.

I guess there should be a sign restriction on `(crit * std_null - crit_power * std_alternative)`

This function needs more unit tests, e.g. against `normal_power_het`

File: statsmodels/stats/power.py (hunk 1 above)
Comment (goxberry) [reply statsmodels/stats/power.py:204]: I'll make this change and note it in the documentation, but I don't think it's appropriate for proportions due to [my last comment](https://github.com/statsmodels/statsmodels/pull/8507#issuecomment-1306322495).

Comment (goxberry) [reply statsmodels/stats/power.py]: Changed in [`b6d264d7`](https://github.com/statsmodels/statsmodels/pull/8507/commits/b6d264d74a29a3883c8d7eef2ec56118d2970564), and added documentation to the `std_alternative` parameter noting this choice of default.

**Comment by josef-pkt** [comment]
The function should also apply to z-tests and arbitrary hypothesis tests based on the normal distribution

I wrote normal_sample_size_one_tail as a shortcut function that can be reused for different hypothesis tests, similar to the other functions like normal_power_het (which I use also for Poisson)

`diff`, `std_null`, and `std_alternative` are parameters of the function, so they can be based on any type of hypothesis test with normal distributed test statistic.
The calling function is responsible to provide the appropriate values for those arguments, e.g. as samplesize_proportions_2indep_onetail does for the comparison of 2 independent Binomial proportions.
(I ran out of patience to add it to the recently added poisson rate functions)

For example normal_sample_size_one_tail should work for z-test for mean difference of 2 independent samples, with appropriate std_null.

Aside: When I was looking at the power functions just now, I got confused again how effect size is defined in my functions. At the time I wrote the power classes, I was following mainly R pwr and G*Power+ packages, without understanding the details different definitions of effect sizes.

**Comment by goxberry** [comment]
Is there a reference for the expression you're using, and the assumptions under which it is valid?

**Comment by josef-pkt** [comment]
> Is there a reference for the expression you're using, and the assumptions under which it is valid?

It's a generic formula, that shows up in articles that I read mainly in the context of proportions, or cases where std_alternative is different from std_null, Binomial proportion, Poisson rates, and similar where variance depends on the mean or mean parameter.
Cases where the variance is independent of the mean, like z-test or t-test for (normal, ...) means, std_alternative is usually equal to std_null.

I agree that more unit tests and other use cases can wait for a different PR.

One part still to check is whether checking `alpha < power` is enough or if we need `(crit * std_null - crit_power * std_alternative) > 0` when std_null and std_alternative differ

**Comment by josef-pkt** [comment]
note 
sigma is in the denominator in https://en.wikipedia.org/wiki/Sample_size_determination#Cumulative_distribution_function
That's the same as multiplying the expression by sigma (= std_null = std_alternative) instead

**Comment by josef-pkt** [comment]
one unrelated fail in quantreg unit tests (in only one environment)
AFAICS no random seed is set
```
  def test_nontrivial_singular_matrix():
        x_one = np.random.random(1000)
...
E       AssertionError: 
E       Not equal to tolerance rtol=0.01, atol=0
```

File: statsmodels/stats/power.py [hunk 2]
```diff
@@ -172,39 +172,49 @@ def normal_sample_size_one_tail(diff, power, alpha, std_null=1.,
...
     crit_power = stats.norm.isf(power)
     crit = stats.norm.isf(alpha)
     n1 = ((crit * std_null - crit_power * std_alternative) / diff)**2
+
+    # Enforce condition that power >= alpha; return np.nan for elements not
+    # satisfying this condition. np.ones_like used to satisfy broadcastability
+    # constraints and to yield a Boolean array with the same shape as n1.
+    alpha_minus_power = (
+        alpha * np.ones_like(std_null) - power * np.ones_like(std_alternative)
+        ) / np.ones_like(diff)
```
Comment (josef-pkt) [inline statsmodels/stats/power.py:216]: diff could be negative, then it could be -/- = +

**Comment by josef-pkt** [comment]
condition is still unclear

currently if alpha == power (and std_alt=std_null), then nobs = 0, i.e. condition holds for any sample size

What should be the returned sample size if power requirements always hold, np.nan or zero?
nobs=zero would mean, power requirement is always satisfied.
however if nobs=0, then std of test statistic is inf, and we compute inf - inf, which would be nan.
another alternative would be to truncate/clip at nobs=1, i.e. nobs=1 means power requirement is satisfied for any nonzero sample size.
 

unequal std:
current samplesize_proportions_2indep_onetail has always std_null > std_alternative, AFAICS. Currently only value=0 is allowed so props under alternative have larger diff than under the null.
I guess this might not be true if `value != 0`. e.g. noninferiority testing (or superiority testing with a margin ?)

So with unequal std, we can still have positive term/nobs even if alpha=power, AFAICS.
(I didn't try the math yet for the derivation of the sample size formula to see where nonnegativity constraint kicks in.)

**Comment by josef-pkt** [comment]
The simplest implementation would be using np.maximum(term, 0) where term is that part that needs to be nonnegative.
That would return nobs = 0 as lower bound, and prevent the bug with incorrect square term

File: statsmodels/stats/power.py (hunk 2 above)
Comment (goxberry) [reply statsmodels/stats/power.py:216]: I'm not sure I understand your comment; the denominator in this expression is an array of all ones with the same shape as `diff` -- the values in `diff` do not change the output of `numpy.ones_like`. The point of using `np.ones_like` is to build `numpy.NDArray[numpy.float64]`s of all ones to satisfy array broadcasting requirements due to the vectorization feature you've requested I preserve.

**Comment by goxberry** [comment]
Attached are screenshots of the textbook pages I have been citing. Worth noting are the differences in sample size expressions between one-population and two-population tests, as well as the differences between sample size expressions for proportion data versus non-proportion data.

CF4536CC-55E4-46F4-8570-A990D11076C7
C200C6E2-0CA3-476E-91E0-05E22CB46B5E
CDD42579-E6F1-401F-99C9-DBA9AC0EB353
417B57D7-AB84-4F82-BF10-2B065F54D65F
809BFA04-C74A-4E73-B726-80C9A7ED9923

**Comment by goxberry** [comment]
The reason `power` cannot be less than `alpha` is because if it were, then the null hypothesis and the alternative hypothesis would have nonempty intersection -- the specific alternative being considered for the false negative condition would not violate the null hypothesis.

**Comment by goxberry** [comment]
> What should be the returned sample size if power requirements always hold, np.nan or zero? nobs=zero would mean, power requirement is always satisfied.

It would mean that the effect size you're using for the power computation doesn't make sense, which is why I chose `np.nan`. In IEEE-754 arithmetic, NaNs propagate through non-Boolean operations on floating-point numbers, which gives users a signal that something aberrant has occurred in their calculations.

All Boolean comparisons between a floating-point number and NaN are false, including `numpy.nan == numpy.nan`, which is why `math.isnan` and `numpy.isnan` exist. Consequently, if an illegal argument has been given to, say, `scipy.stats.norm.isf`, that will return `numpy.nan` when `n1` is computed. The `numpy.where` statement used in the pull request will preserve any `numpy.nan` elements in `n1` that existed upon entry to that line, and can only overwrite elements in `n1` with `numpy.nan` values when the corresponding element of `alpha_minus_power` is positive. If any element of `alpha` or `power` is `math.nan`, then corresponding elements of the Boolean NumPy array `alpha_minus_power` will evaluate to `False`, so those elements of `n1` are not overwritten with `numpy.nan`s in the `numpy.where` statement. But those elements of `n1` should already be `numpy.nan` upon entry to that statement anyway, so those elements of `n1` remain `numpy.nan` upon exiting the function.

**Comment by goxberry** [comment]
Returning a sample size of zero would have to be documented, and that choice would force users to test for that condition, violating Principle of Least Surprise. It would also be inconsistent with what packages like NumPy and SciPy do in many situations (which is return NaN).

**Comment by goxberry** [comment]
> The reason `power `cannot be less than `alpha` is because if it were, then the null hypothesis and the alternative hypothesis would have nonempty intersection -- the specific alternative being considered for the false negative condition would not violate the null hypothesis.

A more precise way to state that condition would be that the effect size used in the power computation would be in the complement of the critical region of the test. I could see returning `numpy.nan` if `power == alpha` because returning zero would be nonsensical; I hadn't given much thought as to whether the critical region should be an open set or a closed set in one dimension.

I ran a quick test using the `pwr` library in R, and setting `power` too close to `alpha` causes a runtime error:

```r
library(pwr)
pwr.t.test(d=0.50, sig.level=0.05, power=0.06, type="one.sample", alternative="two.sided")
```

Result:

```console
Error in uniroot(function(n) eval(p.body) - power, c(2 + 1e-10, 1e+09)) : 
  f() values at end points not of opposite sign
Calls: pwr.t.test -> uniroot
Execution halted
```

which I'd guess is due to numerical integration issues wreaking havoc on Brent's method.

**Comment by josef-pkt** [comment]
Thanks for the reference and copies

The main reason that I have a small preference for returning nobs=0 instead of nobs=nan, is that the power requirement is satisfied (with inequality, power is weakly large than required).
AFAIR, we also have cases in equivalence testing where no solution (of rootfinding) exists, I don't remember the details, but there is an explicit warning for that case.

Also, if we use the one tail approximation for starting values of full rootfinding, then 0 is a valid, finite value, while nans would have to be captured separately.

slightly strange
nobs = 0 does not cause a warning or error in the power computation for proportions_2indep
```
pres = smprop.power_proportions_2indep(diff, prop2, 0, ratio=1, alpha=alphas,
                                value=0, alternative='two-sided',
                                return_results=True)

pres.power, alphas
(0.17699994280370485, 0.2)
```

The results is close nobs=1e-13, i.e. continuous extension to zero.
I guess it's a numerical side effect of how the the power computation is implemented.
Also the lower bound on power (independent of sample size) is smaller than alpha. 

We will have the document the corner behavior in either case whether it's zero or nan.

Another one,
if we return nobs = 0, then users can np.clip nobs to lower bound one, but we cannot clip nan.

**Comment by josef-pkt** [comment]
using maximum with 0, for proportion_2indep
```
powers = np.linspace(1e-4, 0.5, 100)
n = smprop.samplesize_proportions_2indep_onetail(diff, prop2, powers, ratio=1,
                                                 alpha=0.2, value=0)
```

image

**Comment by josef-pkt** [comment]
about "alternative" one versus 2 sided  (I briefly got confused about how this works)

this is handled in `samplesize_proportions_2indep_onetail` which has alternative keyword
```
    if alternative in ['two-sided', '2s']:
        alpha = alpha / 2
```

`power.normal_sample_size_one_tail` was written as helper function and does not have "alternative" keyword
docstring description says
'Use alpha/2 to compute the one tail approximation to the two-sided
    test, i.e. consider only one tail of two-sided test.'

but it's not mentioned again in the alpha parameter explanation.
alpha has the generic, copy-pasted explanation.

**Comment by goxberry** [comment]
> We will have the document the corner behavior in either case whether it's zero or nan.

Agreed. I'll change it to zero. Could possibly be worth a keyword option to toggle behavior re: zero vs nan (or even raising an exception).

**Comment by goxberry** [comment]
@josef-pkt Changed implementation to return `nobs = 0` when `power <= alpha`, and updated the function's docstring to document this behavior in commit [`2de4c9be`](https://github.com/statsmodels/statsmodels/pull/8507/commits/2de4c9be7ce849a47472f18bb436fe491dc20e35).

File: statsmodels/stats/power.py (hunk 2 above)
Comment (josef-pkt) [reply statsmodels/stats/power.py:217]: sorry, I misread

However, we just need (line 210)
`n1 = (np.maximum(crit * std_null - crit_power * std_alternative, 0) / diff)**2`

alpha < power is not the appropriate condition for the term to be nonegative if std_null != std_alternative

Comment (goxberry) [reply statsmodels/stats/power.py:217]: If `crit`, `std_null`, `crit_power`, `std_alternative`, and `diff` are all real, `n1` can never be negative because it is the (elementwise) square of real quantities, so the proposed `numpy.maximum` operation is unnecessary, and will not fix the problem regarding `power <= alpha` for a one-tailed test. The problem is one of well-posedness, not one of algebra or the sign of different quantities.

If we assume that `std_null == std_alternative`, then the situation for a one-tailed test resembles that of this graphic (https://en.wikipedia.org/wiki/Power_of_a_test#/media/File:Statistical_test,_significance_level,_power.png), if we ignore the blue-shaded region in the lower-tail of the blue curve. When `power <= alpha`, the mean of the red curve is less than or equal to the mean of the blue curve, and you cannot have a "false negative" in that situation -- for a one-tailed test, it simply does not make sense because the specific alternative being used in the power analysis will be an alternative consistent with the null hypothesis.

For a two-tailed test, the situation is different because the specific alternative being used in the power analysis will jump from the lower tail to the upper tail, or vice versa.

Comment (goxberry) [reply statsmodels/stats/power.py:217]: If I remember the derivation for the `std_null != std_alternative` case correctly, the test can be posed in terms of the difference of the means of the "null" and "alternative", and a similar sort of situation applies for a one-tailed test. However, the formula I got when I did that derivation was that of page 490 in the screenshots posted in https://github.com/statsmodels/statsmodels/pull/8507#issuecomment-1312146226, not the formula used in the implementation of `normal_sample_size_one_tail`.

Comment (goxberry) [reply statsmodels/stats/power.py:217]: For proportions, the formulas on page 452 of the screenshots posted in https://github.com/statsmodels/statsmodels/pull/8507#issuecomment-1312146226 are different than for non-proportions under the assumption that the variance of a proportion with value `p` is `n * p * (1 - p)`, so the `Z` statistic in that case is not a standard normal variable with zero mean and unit variance. I would have to try to re-derive the result myself to understand better what is happening.

Comment (goxberry) [reply statsmodels/stats/power.py:217]: At this point, I'd be willing to just drop all of the changes related to `alpha` and `power`. My impression is that we disagree about the cause of the problem when `power <= alpha` and when it occurs. The changes regarding multiplication by `None` are less contentious, and we agree on those, so it might be better to land those changes.

re: the `power <= alpha` issue, if that's something that you agree exists and should be fixed, I would be interested in continuing those discussions if we can come to some sort of agreement about what to do, but I fear we are at an impasse.

re: the implementation, I'd also be happy to use `statsmodels.stats.power.normal_sample_size_one_tail` if I understood the assumptions under which it is valid, with references to back that up. Otherwise, I can't tell whether it's appropriate for my application, and can't make a business case to my management that we should use and support this library.

Whatever your decision, I support whatever you think is best. Software maintenance is a lot of work, and it's often under-recognized, particularly when it comes to numerical software, so I want to be conscious of that and want to avoid being a drain on your time or mental health.

File: statsmodels/stats/power.py [hunk 3]
```diff
@@ -172,39 +172,50 @@ def normal_sample_size_one_tail(diff, power, alpha, std_null=1.,
...
-        Hypothesis if the Alternative Hypothesis is true.
+    power : float in interval (alpha,1)
+        power of the test, e.g. 0.8, is one minus the probability of a type II
+        error. Power is the probability that the test correctly rejects the
+        Null Hypothesis if the Alternative Hypothesis is true.
     alpha : float in interval (0,1)
-        significance level, e.g. 0.05, is the probability of a type I
-        error, that is wrong rejections if the Null Hypothesis is true.
+        significance level, e.g. 0.05, is the probability of a type I error,
+        that is wrong rejections if the Null Hypothesis is true.
```
[the remark this answers is not in the document]
Comment (goxberry) [reply statsmodels/stats/power.py]: Added in [`00f27394`](https://github.com/statsmodels/statsmodels/pull/8507/commits/00f27394dbf94de8c028c843ab4e9abd5f8e5aa2).

**Comment by goxberry** [comment]
As far as I can tell, the reason `power` must be at least `alpha` is a consequence of the Neyman-Pearson Lemma regarding uniformly most powerful hypothesis tests for various situations (which includes, under certain assumptions, z-tests). Below is the relevant section of G. G. Roussas, *A Course in Mathematical Statistics*, 2nd Edition, Academic Press, 1997 (pages 329-333):
8522308F-A760-45BB-B035-391CAD0476E6
8A91AC05-3583-4EAB-8AF0-3613081BB00E
8FE45883-1957-4CA8-A0A8-052072AFC5A0
B43B24F6-531E-4C10-A2F8-353FA03818AE
45448E79-B9ED-4EE4-A9A0-AD601E76EA69

An example power function plot demonstrating that its value cannot be less than alpha is shown in the excerpt from J. Gillard, *A First Course in Statistical Inference*, Springer, 2010 (pages 77-79):

818AB219-A340-47CB-B4DC-27C61B773FA7
F9B6EE54-019C-4B9A-AC1E-A72F9EA2034B
0F4DF529-4ADC-4A02-97D6-76ED1DC0A33A

**Comment by josef-pkt** [comment]
Where are just using properties of an existing test, and compute type 1 and type 2 (power) and their relationship to nobs.
Most of our hypothesis tests are not UMP, and I never really looked at Neyman Pearson lemma in detail not used it.

The bug, or one of the bugs, you pointed out at the beginning is that we use a quadrating function where we should only use the positive arm. In the current code we have an incorrect u-shaped function of nobs as function of required power where the initial downward sloping is wrong. That shows a positive nobs, even though it should be nan or zero.

The condition alpha < power is only the lowest point of that quadratic function if std are equal. If std_alt and std_null differ, then the positive arm (nobs > 0) does not start at alpha = power. We have to take into account that those std can differ to find nobs >0 or rule out that we take square of a negative number.

That was all I was trying to explain.

As properties of a specific test, we don't impose alpha < power unless needed for some computation.
The current power functions can be evaluated at points where power < alpha, This happens if we are on the wrong side of a tail the null, It doesn't make much sense to use a hypothesis test like that, but we can compute the type1 and type2 error or power, i.e. rejection probability,  for any point in the parameter space.
for example,
We test that diff =0 against diff > 0
However, we can compute the rejection rate if the true diff is < 0. In that case power < alpha, if rejection rate if diff =0 is true is alpha. 
(It works out in the computation, but we don't really have a reason to do that in practice if we are interested in the "larger" alternative.  I'm using the edge case alpha=power in unit tests quite often.)

**Comment by goxberry** [comment]
> we should only use the positive arm. In the current code we have an incorrect u-shaped function of nobs as function of required power where the initial downward sloping is wrong. That shows a positive nobs, even though it should be nan or zero.

I think I understand that, and will make the suggested change. Thank you for your patience!

**Comment by goxberry** [comment]
Change made re: `np.maximum` in [`16b05325`](https://github.com/statsmodels/statsmodels/pull/8507/commits/16b0532555c60bb280f3554569c3ce270397d995).

**Comment by josef-pkt** [comment]
@goxberry 

I updated the PR in #8544. That will be merged soon
